drivers/serial/uart_miv.c

```diff
@@ -0,0 +1,415 @@
+#define CTRL2_BIT_LENGTH_SHIFT    0
+
+/* Parity enable */
+#define CTRL2_PARITY_EN_OFFSET    0xC
+#define CTRL2_PARITY_EN_MASK      0x02
+#define CTRL2_PARITY_EN_SHIFT     1
+
+/* Odd/even parity configuration */
+#define CTRL2_ODD_EVEN_OFFSET     0xC
+#define CTRL2_ODD_EVEN_MASK       0x04
+#define CTRL2_ODD_EVEN_SHIFT      2
+
+/* Baud value higher 5 bits */
+#define CTRL2_BAUDVALUE_OFFSET    0xC
+#define CTRL2_BAUDVALUE_MASK      0xF8
+#define CTRL2_BAUDVALUE_SHIFT     3
+
+/* Status register */
+#define StatusReg_REG_OFFSET    0x10
+
+#define STATUS_REG_OFFSET       0x10
+
+/* TX ready */
+#define STATUS_TXRDY_OFFSET   0x10
+#define STATUS_TXRDY_MASK     0x01
+#define STATUS_TXRDY_SHIFT    0
+
+/* Receive full - raised even when 1 char arrived */
+#define STATUS_RXFULL_OFFSET   0x10
+#define STATUS_RXFULL_MASK     0x02
+#define STATUS_RXFULL_SHIFT    1
+
+/* Parity error */
+#define STATUS_PARITYERR_OFFSET   0x10
+#define STATUS_PARITYERR_MASK     0x04
+#define STATUS_PARITYERR_SHIFT    2
+
+/* Overflow */
+#define STATUS_OVERFLOW_OFFSET   0x10
+#define STATUS_OVERFLOW_MASK     0x08
+#define STATUS_OVERFLOW_SHIFT    3
+
+/* Frame error */
+#define STATUS_FRAMERR_OFFSET   0x10
+#define STATUS_FRAMERR_MASK     0x10
+#define STATUS_FRAMERR_SHIFT    4
+
+/* Data bits length defines */
+#define DATA_7_BITS     0x00
+#define DATA_8_BITS     0x01
+
+/* Parity defines */
+#define NO_PARITY       0x00
+#define EVEN_PARITY     0x02
+#define ODD_PARITY      0x06
+
+/* Error Status definitions */
+#define UART_PARITY_ERROR    0x01
+#define UART_OVERFLOW_ERROR  0x02
+#define UART_FRAMING_ERROR   0x04
+
+#define BAUDVALUE_LSB ((u16_t)(0x00FF))
+#define BAUDVALUE_MSB ((u16_t)(0xFF00))
+#define BAUDVALUE_SHIFT ((u8_t)(5))
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static struct k_thread rx_thread;
+static K_THREAD_STACK_DEFINE(rx_stack, 512);
+#endif
+
+struct uart_miv_regs_t {
+	u8_t tx;
+	u8_t reserved0[3];
+	u8_t rx;
+	u8_t reserved1[3];
+	u8_t ctrlreg1;
+	u8_t reserved2[3];
+	u8_t ctrlreg2;
+	u8_t reserved3[3];
+	u8_t status;
+};
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+typedef void (*irq_cfg_func_t)(struct device *dev);
+#endif
+
+struct uart_miv_device_config {
+	u32_t       uart_addr;
+	u32_t       sys_clk_freq;
+	u32_t       line_config;
+	u32_t       baud_rate;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	irq_cfg_func_t cfg_func;
+#endif
+};
+
+struct uart_miv_data {
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_t callback;
+#endif
+};
+
+#define DEV_CFG(dev)						\
+	((const struct uart_miv_device_config * const)		\
+	 (dev)->config->config_info)
+#define DEV_UART(dev)						\
+	((struct uart_miv_regs_t *)(DEV_CFG(dev))->uart_addr)
+#define DEV_DATA(dev)						\
+	((struct uart_miv_data * const)(dev)->driver_data)
+
+static unsigned char uart_miv_poll_out(struct device *dev,
+				       unsigned char c)
+{
+	volatile struct uart_miv_regs_t *uart = DEV_UART(dev);
+
+	while (!(uart->status & STATUS_TXRDY_MASK))
```

Comment:
wondering if there could be concurrency problems if two threads are attempting to poll out at the same time.

Reply:
Other serial drivers do not implement any critical section guards in the `poll_*` functions. Actually, the driver functions should not be called directly, but rather through a SYSCALL. The upper Zephyr layers should handle the concurrent accesses.

Reply:
Hi @andrewboie is that an acceptable answer? Or do you have a concrete recommendation analogous to how it's done for some other platform? Reason I'm asking is that nobody's probably going to look at this PR when they see the "Changes requested" status ;)

Reply:
Ah, I see it's @dbkinder's review that's the blocker here, so reiterating the request to him too!